### DISCUSSION TITLE
Upcoming Move to GTK 3 by Default

### DISCUSSION THREAD

**Discussion by orbitalquark** [orbitalquark/textadept#178] [General]
GTK is the cross-platform GUI toolkit Textadept uses. All Textadept binaries have used GTK 2.24 by default, which was released in 2011. That version was end-of-lifed at the end of 2020. The switch to using GTK 3 as a default is inevitable. I have been experimenting with binaries built with GTK 3 on all platforms (Linux, Windows, macOS) and will likely make the switch for the next release.

The user-facing benefits of GTK 3, especially on Windows and macOS is native file chooser dialogs and better support for HiDPI displays. For Linux, Textadept will look like most other GTK apps, many of which have already made the switch to GTK 3. The downside is that on Windows and macOS, GTK 3 apps will look less native than GTK 2 apps (particularly on Windows).

That said, on Linux you will still be able to compile Textadept to use GTK 2 using a "GTK2=1" makefile option. (This is my personal preference). It's just that the default will be GTK 3 moving forward.

If anyone has thoughts or concerns about this, please let me know.

**Comment by paaguti** [comment]
I’ve been compiling for GTK3 on Linux for quite a long time and it works like a breeze:-)

Enviado desde mi iPhone

**Comment by orbitalquark** [comment]
The nightly builds now produce executables linked to GTK 3 on all platforms (Windows, macOS, and Linux). On Windows and macOS the downloads and unzipped archives are larger in size due to the GTK 3 runtime being larger than GTK 2. That is unfortunate.

If you are on Linux and prefer GTK 2, just run `make GTK2=1` when building Textadept from source. For Windows and macOS there is no toggle option, but it is possible to build using GTK 2 by making changes to the Makefile. This is not supported, though.

**Comment by manuelmontenegro** [comment]
Linux user here. The GTK3 version is quite stable, in general. It still has some quirks [1] but I think the advantages (such as HiDPI display support) outweight them. That's why I've been compiling the GTK3 version in the latest releases.

[1] I'll try to create an issue about the "quirks" when I have some time. It is about a misplaced completion popup that shows up when I am working with several monitors, but it's still hard to figure out under which conditions does this happen.

**Reply by manuelmontenegro** [reply]
>  It is about a misplaced completion popup that shows up when I am working with several monitors, but it's still hard to figure out under which conditions does this happen.

Update: It turns out it was a Scintilla-related thing ([Scintilla #2296](https://sourceforge.net/p/scintilla/bugs/2296/) and [Scintilla #2261](https://sourceforge.net/p/scintilla/bugs/2261/)), and was fixed in v5.1.5, so Textadept's nightly (based on Scintilla 5.2.0) works perfectly.

Thanks,
Manuel.

**Comment by paaguti** [comment]
More feedback:
- installed on macOS (Catalina) from nightly: works like a breeze and looks
more than just nice :-)
- installed from locally compiled on Linux (Ubuntu focal): idem...
Thanks a ton!
/PA

On Fri, 4 Mar 2022 at 18:32, Manuel Montenegro ***@***.***>
wrote:

> Linux user here. The GTK3 version is quite stable, in general. It still
> has some quirks [1] but I think the advantages (such as HiDPI display
> support) outweight them. That's why I've been compiling the GTK3 version in
> the latest releases.
>
> [1] I'll try to create an issue about the "quirks" when I have some time.
> It is about a misplaced completion popup that shows up when I am working
> with several monitors, but it's still hard to figure out under which
> conditions does this happen.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/178#discussioncomment-2296750>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4FMXKQQMKH5LJLGQNLU6JCJFANCNFSM5PX2KWBQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

**Comment by orbitalquark** [comment]
Thanks for the positive feedback you two :) I'm glad it's working for more than just me!

**Comment by paaguti** [comment]
Hi,

just a thing... I have seen that I can work with 0xRRGGRR colours but not
with "#RRGGBB" colours in the themes.
Sorry if I'm forgetting the evolution and you justified the move.... but
could we have the "#RRGGBB" back.
I 'm trying to include TextAdept in ThemeCreator and having the
(emacs-compatible) "#RRGGBB" format would make this inclusion a breeze.

Best, /PA

On Sun, 6 Mar 2022 at 15:11, orbitalquark ***@***.***> wrote:

> Thanks for the positive feedback you two :) I'm glad it's working for more
> than just me!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/178#discussioncomment-2303431>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4E2J64EU4PJ7HWRLJTU6S4HNANCNFSM5PX2KWBQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

**Comment by paaguti** [comment]
Discard the last message... I need a couple cups of coffe more ... :-(

Best, /PA

On Mon, 7 Mar 2022 at 08:28, Pedro Andres Aranda Gutierrez <
***@***.***> wrote:

> Hi,
>
> just a thing... I have seen that I can work with 0xRRGGRR colours but not
> with "#RRGGBB" colours in the themes.
> Sorry if I'm forgetting the evolution and you justified the move.... but
> could we have the "#RRGGBB" back.
> I 'm trying to include TextAdept in ThemeCreator and having the
> (emacs-compatible) "#RRGGBB" format would make this inclusion a breeze.
>
> Best, /PA
>
> On Sun, 6 Mar 2022 at 15:11, orbitalquark ***@***.***>
> wrote:
>
>> Thanks for the positive feedback you two :) I'm glad it's working for
>> more than just me!
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/orbitalquark/textadept/discussions/178#discussioncomment-2303431>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAZKU4E2J64EU4PJ7HWRLJTU6S4HNANCNFSM5PX2KWBQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>
>
> --
> Fragen sind nicht da um beantwortet zu werden,
> Fragen sind da um gestellt zu werden
> Georg Kreisler
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

**Comment by paaguti** [comment]
BTW, to make up for the last two messages:

I have a laptop with a Ubuntu 22.04pre and just compiled TextAdept on it.
Works wonderfully :-)

Best,/PA

On Mon, 7 Mar 2022 at 08:30, Pedro Andres Aranda Gutierrez <
***@***.***> wrote:

> Discard the last message... I need a couple cups of coffe more ... :-(
>
> Best, /PA
>
> On Mon, 7 Mar 2022 at 08:28, Pedro Andres Aranda Gutierrez <
> ***@***.***> wrote:
>
>> Hi,
>>
>> just a thing... I have seen that I can work with 0xRRGGRR colours but not
>> with "#RRGGBB" colours in the themes.
>> Sorry if I'm forgetting the evolution and you justified the move.... but
>> could we have the "#RRGGBB" back.
>> I 'm trying to include TextAdept in ThemeCreator and having the
>> (emacs-compatible) "#RRGGBB" format would make this inclusion a breeze.
>>
>> Best, /PA
>>
>> On Sun, 6 Mar 2022 at 15:11, orbitalquark ***@***.***>
>> wrote:
>>
>>> Thanks for the positive feedback you two :) I'm glad it's working for
>>> more than just me!
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/orbitalquark/textadept/discussions/178#discussioncomment-2303431>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AAZKU4E2J64EU4PJ7HWRLJTU6S4HNANCNFSM5PX2KWBQ>
>>> .
>>> Triage notifications on the go with GitHub Mobile for iOS
>>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>>> or Android
>>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>>
>>> You are receiving this because you commented.Message ID:
>>> ***@***.***
>>> com>
>>>
>>
>>
>> --
>> Fragen sind nicht da um beantwortet zu werden,
>> Fragen sind da um gestellt zu werden
>> Georg Kreisler
>>
>
>
> --
> Fragen sind nicht da um beantwortet zu werden,
> Fragen sind da um gestellt zu werden
> Georg Kreisler
>


-- 
Fragen sind nicht da um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

**Reply by snoopy** [reply]
Seems like the statusbar font size was decreased in gtk3. Any way to increase it again? I tried looking in the gtk.css file but couldn't find the right setting.

**Comment by orbitalquark** [comment]
If you put this at the end of gtk.css:

  * { font: 15px sans; }

That seems to increase the font size for me.

**Reply by snoopy** [reply]
That works for me, thanks.

Is there a way to only change the Textadept styling? Placing a gtk.css in your home folder seems to affect all GTK3 applications.

**Reply by orbitalquark** [reply]
I am not very familiar with GTK 3 and per-application styling. Some quick searches have indicated that you might be able to use a GTK_THEME environment variable. On Linux you may be able to copy an existing theme into a new one, and run Textadept with GTK_THEME pointing to your custom theme.

On Windows and macOS you'll probably end up having to edit the gtk.css files manually.

Sorry I cannot be of more help :(

**Comment by gabdub** [comment]
Are you planning to expose the compiled GTK version to lua?
I noticed that the names of the icons change with the version.
```
icon = 'gtk-dialog-question'  --GTK2
icon = 'dialog-question'  --GTK3
```

**Reply by orbitalquark** [reply]
No I don't. That icon name, as well as the other dialog icon names used, work with both GTK 2 and GTK 3.

**Reply by gabdub** [reply]
OK,  thanks.

**Reply by gabdub** [reply]
I don't know why the compiled version doesn't show the dialog icons (Windows 64 - GTK 3.24.31).

The first image is from textadept.exe nightly 2022-3-14 (20.2MB):
![dlg no icon](https://user-images.githubusercontent.com/11639609/158172658-1d73d369-5a4d-4125-aa06-0704b5cb9920.png)
The second one is from the compiled version (2.5MB) using a VM with Ubuntu 20.04.4 with libgtk3-dev installed after running:
```
make win-deps
make win

```
Everything seems to be working except this.
Any ideas?

**Reply by orbitalquark** [reply]
You need to use `make win-deps NIGHTLY=1` to grab the nightly version of gtdialog that uses the desktop icon naming specification.

**Reply by gabdub** [reply]
You're right, sorry for the inconvenience.

**Reply by orbitalquark** [reply]
No problem. I've done that a lot :)

**Comment by orbitalquark** [comment]
No I don't. That icon name, as well as the other dialog icon names used, work with both GTK 2 and GTK 3.
